sky/backends/backend_utils.py

```diff
@@ -1775,32 +1777,50 @@ def _update_cluster_status_no_lock(
         if external_ips is None or len(external_ips) == 0:
             raise exceptions.FetchIPError(
                 reason=exceptions.FetchIPError.Reason.HEAD)
-        if handle.launched_nodes == 1:
-            # Check the ray cluster status. We have to check it for single node
-            # case, since the get_node_ips() does not require ray cluster to be
-            # running.
-            ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml)
-            runner = command_runner.SSHCommandRunner(external_ips[0],
-                                                     **ssh_credentials)
-            returncode = runner.run('ray status', stream_logs=False)
-            if returncode:
-                raise exceptions.FetchIPError(
-                    reason=exceptions.FetchIPError.Reason.HEAD)
+        # Check if ray cluster status is healthy.
+        ssh_credentials = ssh_credential_from_yaml(handle.cluster_yaml)
+        runner = command_runner.SSHCommandRunner(external_ips[0],
+                                                 **ssh_credentials)
+        returncode = runner.run('ray status', stream_logs=False)
+        if returncode:
+            raise exceptions.FetchIPError(
+                reason=exceptions.FetchIPError.Reason.HEAD)
+        ray_cluster_up = True
+        # For non-spot clusters:
         # If we get node ips correctly, the cluster is UP. It is safe to
         # set the status to UP, as the `handle.external_ips` function uses ray
         # to fetch IPs and starting ray is the final step of sky launch.
+        # For spot clusters, the above can be unsafe because the Ray cluster
+        # may remain healty for a while before the cloud completely
+        # terminates the VMs.
+        # Additionally, we query the VM state from the cloud provider.
```

Comment:
Is the fix for checking the `ray status` for multiple nodes already enough? The previous problem might because the worker VM is preempted, but the IP can still be got with `ray get-node-ips`?

Did the problem happened for a user with a single node or multiple nodes?

Reply:
The problem happened to a user with a single node (tpu-v2-8) I believe